libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription_container.c

```diff
@@ -0,0 +1,358 @@
+            }
+        }
+        else
+        {
+            /* Check for wildcards. */
+            if( pTopicFilter[ filterIndex ] == '+' )
+            {
+                /* Move topic name index to the end of the current level.
+                 * This is identified by '/'. */
+                while( nameIndex < topicNameLength && pTopicName[ nameIndex ] != '/' )
+                {
+                    nameIndex++;
+                }
+
+                /* Increment filter index to skip '/'. */
+                filterIndex++;
+                continue;
+            }
+            else if( pTopicFilter[ filterIndex ] == '#' )
+            {
+                /* Subsequent characters don't need to be checked if the for the
+                 * multi-level wildcard. */
+                IOT_SET_AND_GOTO_CLEANUP( true );
+            }
+            else
+            {
+                /* Any character mismatch other than '+' or '#' means the topic
+                 * name does not match the topic filter. */
+                IOT_SET_AND_GOTO_CLEANUP( false );
+            }
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    /* If the end of both strings has been reached, they match. */
+    if( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) )
+    {
+        IOT_SET_AND_GOTO_CLEANUP( true );
+    }
+
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
+}
+
+/*-----------------------------------------------------------*/
+
+int8_t IotMqtt_InsertSubscription( _mqttSubscription_t * pSubscriptionArray )
+{
+    /* The shim supports only upto 256 subscriptions as the implementation uses 8 bit index. */
+    int8_t index = 0;
+
+    IotMqtt_Assert( pSubscriptionArray != NULL );
+
+    /* Finding the free index in subscription array to insert the subscription. */
+    while( index < MAX_NO_OF_MQTT_SUBSCRIPTIONS )
+    {
+        /* Using topicFilterLength as a unique parameter to check whether index is free or not. */
+        if( pSubscriptionArray[ index ].topicFilterLength == 0 )
+        {
+            break;
+        }
+
+        index++;
+    }
+
+    if( index == MAX_NO_OF_MQTT_SUBSCRIPTIONS )
+    {
+        index = -1;
+    }
+
+    return index;
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotMqtt_RemoveSubscription( _mqttSubscription_t * pSubscriptionArray,
+                                 int8_t deleteIndex )
+{
+    bool status = false;
+
+    IotMqtt_Assert( pSubscriptionArray != NULL );
+    IotMqtt_Assert( deleteIndex > -1 && deleteIndex < MAX_NO_OF_MQTT_SUBSCRIPTIONS );
+
+    /* Remove the subscription from the subscription array. */
+    if( deleteIndex != -1 )
+    {
+        /* Using topicFilterLength as a unique parameter to free index and make it available for other subscriptions. */
+        pSubscriptionArray[ deleteIndex ].topicFilterLength = 0;
+        status = true;
+    }
+    else
+    {
+        status = false;
```

Comment:
Do we need status update to be false here?

Reply:
Addressed